tests/testing_utils/fixtures.py

```diff
@@ -0,0 +1,194 @@
+
+
+@pytest.fixture()
+def mock_ctx(mock_cursor):
+    class _MockConnectionCtx(mock.MagicMock):
+        def __init__(self, cursor=None, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.queries: List[str] = []
+            self.cs = cursor
+
+        def get_query(self):
+            return "\n".join(self.queries)
+
+        def get_queries(self):
+            return self.queries
+
+        @property
+        def warehouse(self):
+            return "MockWarehouse"
+
+        @property
+        def database(self):
+            return "MockDatabase"
+
+        @property
+        def schema(self):
+            return "MockSchema"
+
+        @property
+        def role(self):
+            return "MockRole"
+
+        @property
+        def host(self):
+            return "account.test.region.aws.snowflakecomputing.com"
+
+        def execute_string(self, query: str):
+            self.queries.append(query)
+            if self.cs:
+                return (self.cs,)
+            else:
+                return (mock_cursor(["row"], []),)
+
+        def execute_stream(self, query: StringIO):
+            return self.execute_string(query.read())
+
+    return lambda cursor=None: _MockConnectionCtx(cursor)
+
+
+@pytest.fixture
+def mock_cursor():
+    class MockResultMetadata(NamedTuple):
+        name: str
+
+    class _MockCursor(SnowflakeCursor):
+        def __init__(self, rows: List[tuple], columns: List[str]):
+            super().__init__(mock.Mock())
+            self._rows = rows
+            self._columns = [MockResultMetadata(c) for c in columns]
+
+        def fetchone(self):
+            return self.fetchall()
+
+        def fetchall(self):
+            return self._rows
+
+        @property
+        def description(self):
+            yield from self._columns
+
+        @classmethod
+        def from_input(cls, rows, columns):
+            return cls(rows, columns)
+
+    return _MockCursor.from_input
+
+
+@pytest.fixture
+def other_directory() -> Generator:
+    tmp_dir = tempfile.TemporaryDirectory()
+    yield tmp_dir.name
+    tmp_dir.cleanup()
+
+
+@pytest.fixture
+def other_directory_with_chdir(other_directory: str) -> Generator:
+    initial_dir = os.getcwd()
+    os.chdir(other_directory)
+    yield other_directory
+    os.chdir(initial_dir)
+
+
+@pytest.fixture
+def package_file():
+    with tempfile.TemporaryDirectory() as tmp:
+        yield create_named_file("app.zip", tmp, [])
+
+
+@pytest.fixture(scope="function")
+def runner(test_snowcli_config):
+    from snowcli.cli.app import app
+
+    return SnowCLIRunner(app, test_snowcli_config)
+
+
+@pytest.fixture
+def temp_dir():
+    initial_dir = os.getcwd()
+    tmp = tempfile.TemporaryDirectory()
+    os.chdir(tmp.name)
+    yield tmp.name
+    os.chdir(initial_dir)
+    tmp.cleanup()
+
+
+@pytest.fixture
+def temp_directory_for_app_zip(temp_dir) -> Generator:
+    temp_dir = tempfile.TemporaryDirectory(dir=temp_dir)
+    yield temp_dir.name
```

Comment:
What is the difference between `temp_dir`, `temp_directory_for_app_zip` and `other_directory_with_chdir`?

Reply:
`temp_dir` is basic directory for all tests, `temp_directory_for_app_zip` is a subdir in `temp_dir`, and `other_directory_with_chdir` is a separate directory, for tests, that need two directories in different paths (but it seems most of them use only `other_directory`, so i'll delete it)